compiler/bindings/python/iree/compiler/tools/onnx.py

```diff
@@ -0,0 +1,171 @@
+from ..extras import onnx_importer
+from ..ir import Context, StringAttr
+from .binaries import invoke_pipeline
+from .core import build_compile_command_line
+from ...runtime import Config, VmInstance, VmModule, load_vm_module
+
+__all__ = [
+    "compile_saved_model",
+    "compile_model"
+]
+
+
+logger = logging.getLogger(__name__)
+
+
+def load_onnx_model(raw_model: onnx.ModelProto,
+                    data_prop: bool = True,
+                    data_dir: IO[bytes] | str | os.PathLike | None = None):
+    # Load the model, with possible external data coming from the default
+    # location, or the location specified on the command line.
+    if data_dir:
+        onnx.load_external_data_for_model(raw_model, str(data_dir))
+
+    # Do shape inference two ways.  First, attempt in-memory to avoid redundant
+    # loading and the need for writing a temporary file somewhere.  If that
+    # fails, typically because of the 2 GB protobuf size limit, try again via
+    # files.  See
+    # https://onnx.ai/onnx/repo-docs/PythonAPIOverview.html#shape-inference-a-large-onnx-model-2gb
+    # for details about the file-based technique.
+
+    # Run the checker to test whether the file is above the threshold for
+    # in-memory shape inference.  If not, go ahead and do the shape inference.
+    try:
+        onnx.checker.check_model(raw_model)
+        inferred_model = onnx.shape_inference.infer_shapes(
+            raw_model, data_prop=data_prop
+        )
+        return inferred_model
+    except ValueError:
+        pass
+
+    # Model is too big for in-memory inference: do file-based shape inference
+    # to a temp file.
+    # Make a temp dir for all the temp files we'll be generating as a side
+    # effect of infering shapes. For now, the only file is a new .onnx holding
+    # the revised model with shapes.
+    with tempfile.TemporaryDirectory() as temp_dir_name:
+        temp_dir_path = Path(temp_dir_name)
+        temp_input_file = temp_dir_path / "temp.onnx"
+        temp_inferred_file = temp_dir_path / "temp-inferred.onnx"
+        onnx.save(raw_model, temp_input_file)
+        onnx.shape_inference.infer_shapes_path(
+            temp_input_file, temp_inferred_file, data_prop=data_prop
+        )
+
+        # Load the temp file and the external data.
+        inferred_model = onnx.load(
+            temp_inferred_file, load_external_data=False)
+        data_dir = Path(temp_input_file if data_dir is None else data_dir)
+        onnx.load_external_data_for_model(inferred_model, str(data_dir))
+
+        return inferred_model
+
+
+@dataclass
+class ImportOptions(CompilerOptions):
+    input_type: Union[InputType, str] = InputType.ONNX
+    entry_point_name: Optional[str] = None
+    module_name: Optional[str] = None
+    import_only: bool = False
+    save_temp_iree_input: Optional[str] = None
+    verify_module: bool = False
+    use_bytecode: bool = False
+
+
+def compile_saved_model(model_path: IO[bytes] | str | os.PathLike, **kwargs):
```

Comment:
It would be helpful to have a return type like `Optional[???]` (does this return `IO[bytes]` for the compiled vmfb?)

Reply:
the return type should be `None | str | IO[bytes]`


```python
if output_file:
  return None
else:
  if import_only:
    if use_bytecode:
      return IO[bytes]  // mlir bytecode
    else:
      return str // mlir
  else:
    return IO[bytes] // vmfb
```